rfq/negotiator_test.go

```diff
@@ -0,0 +1,349 @@
+
+	// QuoteRespStatus is the expected status of the quote check.
+	QuoteRespStatus fn.Option[QuoteRespStatus]
+}
+
+// assertIncomingBuyAcceptTestCase asserts the handling of an incoming buy
+// accept message for a test case.
+func assertIncomingBuyAcceptTestCase(
+	t *testing.T, tc testCaseIncomingBuyAccept) {
+
+	// Create a mock price oracle.
+	mockPriceOracle := &MockPriceOracle{}
+
+	// Register an expected call and response for price oracle method
+	// QueryBidPrice.
+	mockPriceOracle.On(
+		"QueryBidPrice", mock.Anything, mock.Anything,
+		mock.Anything, mock.Anything, mock.Anything,
+	).Return(
+		&OracleResponse{
+			AssetRate: tc.PriceOracleBidPrice,
+		}, nil,
+	)
+
+	// Define buy request and buy accept messages.
+	var (
+		assetSpecifier = asset.NewSpecifierFromId(asset.ID{1, 2, 3})
+		peerID         = route.Vertex{1, 2, 3}
+		msgID          = rfqmsg.ID{1, 2, 3}
+	)
+
+	buyRequest := rfqmsg.BuyRequest{
+		Peer:           peerID,
+		ID:             msgID,
+		AssetSpecifier: assetSpecifier,
+		AssetMaxAmt:    1000,
+		AssetRateHint:  fn.None[rfqmsg.AssetRate](),
+	}
+
+	buyAccept := rfqmsg.BuyAccept{
+		Request:   buyRequest,
+		AssetRate: tc.IncomingBuyAcceptRate,
+	}
+
+	// Create the negotiator.
+	errChan := make(chan error, 1)
+	negotiator, err := NewNegotiator(NegotiatorCfg{
+		PriceOracle:             mockPriceOracle,
+		OutgoingMessages:        make(chan rfqmsg.OutgoingMsg, 1),
+		AcceptPriceDeviationPpm: tc.AcceptPriceDeviationPpm,
+		ErrChan:                 errChan,
+	})
+	require.NoError(t, err)
+
+	// Define the finalise callback function.
+	finalise := func(
+		msg rfqmsg.BuyAccept, event fn.Option[InvalidQuoteRespEvent],
+	) {
+		// If the actual event is none and the expected status is none,
+		// then we don't need to check anything.
+		if event.IsNone() && tc.QuoteRespStatus.IsNone() {
+			return
+		}
+
+		require.Equal(t, tc.QuoteRespStatus.IsSome(), event.IsSome())
+
+		// Extract the actual event status.
+		var actualEventStatus QuoteRespStatus
+		event.WhenSome(func(e InvalidQuoteRespEvent) {
+			actualEventStatus = e.Status
+		})
+
+		// Extract the expected event status.
+		var expectedStatus QuoteRespStatus
+		tc.QuoteRespStatus.WhenSome(func(e QuoteRespStatus) {
+			expectedStatus = e
+		})
+
+		// Ensure that the actual and expected event statuses are equal.
+		require.Equal(t, expectedStatus, actualEventStatus)
+	}
+
+	// Handle the incoming buy accept message.
+	negotiator.HandleIncomingBuyAccept(buyAccept, finalise)
+
+	// Check that there are no errors.
+	select {
+	case err := <-errChan:
+		t.Fatalf("unexpected error: %v", err)
+	default:
+	}
+
+	// Wait for the negotiator to finish.
+	negotiator.Wg.Wait()
+}
+
+// TestHandleIncomingBuyAccept tests the handling of an incoming buy accept
+// message.
+func TestHandleIncomingBuyAccept(t *testing.T) {
+	defaultQuoteExpiry := time.Now().Add(time.Hour)
+
+	testCases := []testCaseIncomingBuyAccept{
+		{
+			Name: "accept price within bounds",
```

Comment:
Should we also add a test case where the price is within the bounds, but just barely? Instead of only being exactly the same? Here and in the test above.

Reply:
Done.

But see this PR for an explanation of the numbers: https://github.com/lightninglabs/taproot-assets/pull/1377